### ISSUE TITLE
Order of columns in the transaction list is not saved

### ISSUE THREAD

**Issue by tactilis** [moneymanagerex/moneymanagerex#4160]
MMEX version:
 - [x] 1.5.12
 - [ ] 1.6.x

Operating System:
 - [x] Windows
 - [ ] Mac OSX NOT TESTED
 - [ ] Linux NOT TESTED

# Description of the bug

If a user changes the order of columns in the transaction list, then closes and re-opens MMEX, the user selected order is not remembered. The default column order is always displayed.

# Reproduction

Is the bug reproducible? 
 - [x] Always 
 - [ ] Randomly 
 - [ ] Happened only once

Reproduction steps:

1. Drag any column in the transaction list to a new position.
2. Close MMEX.
3. Re-open MMEX.
4. Note that the columns are not displayed in the order set in Step 1.

Expected result:

The user selected column order should be saved and then re-applied when MMEX is re-opened.

Actual result:

The column order always reverts to the default.

# Additional information

Note that:
1. If you right click a column heading and Hide/Show columns, then that user selection _is_ saved and re-applied when MMEX is re-opened.
2. If you change a column width, then that user selection _is_ saved and re-applied when MMEX is re-opened.

**Comment by tactilis** [comment]
@whalley 
> You should think yourself lucky to be able to re-order columns. This functionality only works on Windows and is not available on MacOS or Linux :-)

Smiley noted.

However, if the user is able to set the column order, then it really ought to be remembered across re-starts of MMEX.

Is the lack of ability to re-order columns on MacOS and Linux a platform limitation, or an OS-specific wxWidgets limitation?

**Comment by whalley** [comment]
Its more a wxWidgets limitation - https://docs.wxwidgets.org/trunk/classwx_list_ctrl.html#a2a499df49ebcefb3451a44146f76457d

I agree that if re-ordering is supported then the order should be saved.

**Comment by whalley** [comment]
@moritzhoewer The lack of support for this for Linux and MacOS is with the wxWidgets team https://github.com/wxWidgets/wxWidgets/issues/19007 but has been there for a while. I don't see this being fixed any time soon.

**Comment by erottensteiner** [comment]
Is there any chance to fix this bug?

**Comment by robvatt** [comment]
Hi I'm testin MMEX in order to decide what software to use in personal finance.
Desktop environment Windows 7 MMEX 1.7.0
The problem remains: it seems to be impossible to save changes done in columns order; during a session I can modify column order, hide columns resize columns but the only things to be saved are size and column showed, not column order

**Comment by whalley** [comment]
Will up priority to the v1.7.2 release.

**Comment by whalley** [comment]
Perhaps to avoid the issue of not being able to re-order columns via native wxWidgets we could add option to do this outside the controls on these platforms. Will investigate.

**Comment by n-stein** [comment]
> we could add option to do this outside the controls on these platforms. Will investigate.

It would be fairly easy to make a simple column order dialog that can be called from the View menu and store the desired order in an array in INFOTABLE_V1. That would make it platform agnostic.

**Comment by WebProd33** [comment]
For me (using Windows 11), I can change the size of columns and re order them, except for the "Tags" one. The "Tags" one always go back to right after the "Category" one.

**Comment by jensb** [comment]
Please add a way to reorder the transaction columns. This was the first thing I tried when starting MMEX again after a long period of not looking at it, and it still does not work.
Thank you!

**Comment by jensb** [comment]
@n-stein, can you advise me where to look in the source to create such a dialog? Then I can try and have a go at it.

**Comment by n-stein** [comment]
Where to create it? The cleanest method is to simply add a new .h and .cpp file for your new dialog class. See one of the many other dialog classes for reference, e.g. tagdialog.h/.cpp. You can add menu paths that call your dialog when clicked in mmframe.cpp or modify one of the optionsettings classes to call it from the options dialog.

Where to store the results? A few options here:
1. Store the result as an array in INFOTABLE_V1. This will apply per database meaning two databases can have different column orders when viewed in the same MMEX instance. Downside is if you have multiple databases and you want a specific column order you have to set it manually in each database.
2. Store the result as an array in SETTINGS_V1. This will apply program wide. Every database you open will have the same column order. Downside is if multiple people with different databases share a computer, they can't each have their own order.
3. A combination of 1 and 2. Always store the database setting, but give the user a checkbox to set the order system-wide. If they choose system-wide, update both SETTINGS_V1 and INFOTABLE_V1, otherwise update only INFOTABLE_V1. Dragging a column in the table should only update INFOTABLE_V1. When applying the column order the INFOTABLE_V1 order (if it exists) should override the SETTINGS_V1 order.

How to use them when building the transaction list? Look in mmcheckingpanel.cpp at function `CreateControls` and you'll see the call to `createColumns`. Your choice whether you modify the order before or after that step. You'll also want to capture column drag events in mmchecking_list.cpp to update the INFOTABLE_V1 order.

**Comment by Lanna253** [comment]
wow took me this long to learn just a little how use my acc i apologize gor the inconvenience wasn't intended i honestly didn't know how navigate the acc

**Comment by jensb** [comment]
@n-stein, thanks for the initial pointers!
I forked mmex at jensb/moneymanagerex and will submit a PR as soon as I get anything that works, for review.

**Comment by jensb** [comment]
Status: I have a working stub. :grin: 
![image](https://github.com/user-attachments/assets/2b5c0d3c-f175-4c1f-8d2c-9ca8cafbe178)

I'll use a row in INFOTABLE_V1 as storage at first, since columns may differ depending on what kind of budget you are managing. Later I may update it to let users choose.

**Comment by jensb** [comment]
I've got it, except for the very last step: actually applying the newly sorted list of column names to the list control.
There are multiple things to consider here - column names vs. index, column widths, and (possibly) i18n.

See https://github.com/moneymanagerex/moneymanagerex/pull/6821

Basically `mmcheckingpanel.cpp` now looks like this:
![image](https://github.com/user-attachments/assets/5b98219e-ec0e-425f-889e-45b521602d2a)

How would you go about applying the new sort order to the transaction list?
I was hoping for an in-place reordering but it doesn't seem to be that simple. May need some support here...

**Comment by jensb** [comment]
I experimented with an alternative solution which (IMHO) is more compact: reordering via context menu.
![grafik](https://github.com/user-attachments/assets/3a121a94-e7e9-4ddd-93d8-16f28556ab05)

Storing the sort order in the settings would be the same, this would just make the dialog redundant (2 fewer files).
Which one would you prefer?

**Comment by jensb** [comment]
Can any of the developers please comment on my PR?
I'd like this to be included in mainline, and if it doesn't meet your guidelines, I'll improve it.

Thanks!

**Comment by whalley** [comment]
Will have a look. Have merged the PR allowing users to test and feedback.

**Comment by whalley** [comment]
@jensb  Initial results of testing on MacOS

1. The "Move column left" / Move column right" does not appear to work
2. The "Column Order Manager" works but requires MMEX to be restarted before the change takes effect.

I will investigate further.

**Comment by jensb** [comment]
1. True. It's a stub so far. I can add functionality if you like it.
2. You have to change the view once, ie. go to another account or transactions view and then back. I haven't found out how to force a refresh / redisplay of the transactions view, this would be needed when destroying the dialog window.

**Comment by n-stein** [comment]
> I haven't found out how to force a refresh / redisplay of the transactions view

Something like this 8bf845af11f3ab09b053f76a5fa34de7777e1945 works

**Comment by n-stein** [comment]
I think we should also save the column order in the checking panel destructor. That way if Windows users use the native header-drag functionality to move the columns left/right the order is saved.

**Comment by n-stein** [comment]
Reordering does not work in languages other than English. Need to wrap the check in `wxGetTranslation`.

![image](https://github.com/user-attachments/assets/03a9b371-696d-4749-b3f5-cf12b5d17e52)

**Comment by jensb** [comment]
Thought about something like this but I got confused about m_real_colums[x], at first I thought the enums mean that the column order is hardcoded.
I will take a look at this. Thanks for the pointers!

**Comment by jensb** [comment]
I have this almost working with IDs instead of column names, which should also work if you change the language.
But the dialog now also shows just the IDs. How do I get from an ID to the (translatable) label?

**Comment by jensb** [comment]
I have this almost working with IDs instead of column names, which should also work if you change the language.
But the dialog now also shows just the IDs. How do I get from an ID to the (translatable) label?

**Comment by n-stein** [comment]
> How do I get from an ID to the (translatable) label to use in the dialog?

I think if you have an instance of TransactionListCtrl you would find the index of the COL enum in m_real_columns and look up the HEADER of the same index in m_columns. But without an instance of TransactionListCtrl there isn't a simple way as far as I can see.

You could implement a reverse-mapping in mmchecking_list using a `static std::map<int, wxString>` and map the COL enums to their translated names. That way you can reference them from anywhere.

Something like this 2096c30cf9acc1b4d14881e47a0ecc8a2a9368ad

Then just use `TransactionListCtrl::COLUMN_NAMES[ <column enum> ]` or `TransactionListCtrl::COLUMN_NAMES.at( <column enum> )` to retrieve the translated column name for standard columns. If you have an enum in the range COL_UDFC01 to COL_UDFC05 then you have to do `Model_CustomField::getUDFCName(Model_Attachment::reftype_desc(Model_Attachment::TRANSACTION), TransactionListCtrl::COLUMN_NAMES[ <column enum> ])`

**Comment by jensb** [comment]
Here's a thought: if we skip the dialog entirely and just use the "Move column left/right" context menus in the column headers, *and* include a clean (and fast) way of reloading the current TransactionListCtrl instance so users see the results at once, then we can store column IDs in the settings without having to translate them to strings at all, anywhere. And it's (possibly) less code.
Unfortunately, https://github.com/moneymanagerex/moneymanagerex/commit/8bf845af11f3ab09b053f76a5fa34de7777e1945 resets the column order entirely, so does not work in this case.

What do you think?

**Comment by n-stein** [comment]
> Unfortunately, [8bf845a](https://github.com/moneymanagerex/moneymanagerex/commit/8bf845af11f3ab09b053f76a5fa34de7777e1945) resets the column order entirely, so does not work in this case.

Hmm, I think it only resets the column order because it is still using names instead of enums. If you adjust it to compare enums to m_real_columns instead of headers to m_columns it should work the same.

But regardless it should be fairly simple to just swap the column order and refresh the view with something like this:

```
void mmListCtrl::OnHeaderMove(wxCommandEvent& WXUNUSED(event), int direction)
{
    if (0 <= m_ColumnHeaderNbr + direction && m_columns.size() > m_ColumnHeaderNbr + direction)
    {
        wxLogDebug("Moving column %d (%s) %d", m_ColumnHeaderNbr, m_columns[m_ColumnHeaderNbr].HEADER.c_str(), direction);

        // swap order of column data
        std::swap(m_real_columns[m_ColumnHeaderNbr + direction], m_real_columns[m_ColumnHeaderNbr]);
        std::swap(m_columns[m_ColumnHeaderNbr + direction], m_columns[m_ColumnHeaderNbr]);

        // swap column headers & widths
        wxListItem col1, col2;
        col1.SetText(m_columns[m_ColumnHeaderNbr].HEADER);
        col1.SetWidth(m_columns[m_ColumnHeaderNbr].WIDTH);
        col2.SetText(m_columns[m_ColumnHeaderNbr + direction].HEADER);
        col2.SetWidth(m_columns[m_ColumnHeaderNbr + direction].WIDTH);
        SetColumn(m_ColumnHeaderNbr, col1);
        SetColumn(m_ColumnHeaderNbr + direction, col2);

        // refresh the list view
        Refresh();
    }
}
```

**Comment by jensb** [comment]
I am starting to like this solution more than the dialog.
Less code and not having to translate between i18n strings and column IDs.

I played with the code above, it did not transfer column alignment when moving columns, I fixed this.
Now thinking about how to best store this information. Probably it is best to create a SETTINGS entry for each type of panel (transactions, bills_deposits, assets, ...) since each of these have different columns.

Will commit and push when it works. :-)
